simulator/infra/cache/cache_tag_array.h

```diff
@@ -1,82 +1,127 @@
-public:
+    protected:
+        CacheTagArrayCheck( uint32 size_in_bytes,
+                            uint32 ways,
+                            uint32 line_size,
+                            uint32 addr_size_in_bits);
+
+    public:
         const uint32 size_in_bytes;
-        const uint32 ways;
+        const uint32 number_of_ways;
         const uint32 line_size;
         const uint32 addr_size_in_bits;
 };
 
-class CacheTagArray : public CacheTagArrayCheck
-{
-    private:
-        /* The set (line) of the cache. */
-        struct CacheTag
-        {
-            Addr line = 0; // data
-            bool is_valid = false; // validity
-        };
-
-        const uint32 num_sets;
-        const Addr   addr_mask;
 
-        std::vector<std::vector<CacheTag>> array; // array of tags
-        LRUInfo lru; // LRU algorithm module
 
+class CacheTagArray : public CacheTagArrayCheck
+{
     public:
-        CacheTagArray( uint32 size_in_bytes,
+        CacheTagArray( uint32 size_in_bytes, 
                        uint32 ways,
                        uint32 line_size = 4,
-                       uint32 addr_size_in_bits = 32);
+                       uint32 addr_size_in_bits = 32)
+
+            : CacheTagArrayCheck( size_in_bytes, 
+                                  ways,
+                                  line_size,
+                                  addr_size_in_bits)
+            , number_of_sets( size_in_bytes / ( ways * line_size))
+            , addr_mask( bitmask<Addr>( addr_size_in_bits))
+            , ways_to_tags( number_of_sets, 
+                            std::vector<std::pair<bool, Addr>>( number_of_ways,
+                                                                std::pair<bool, Addr>( false, 0)))
+            , data( number_of_sets, std::unordered_map<Addr, uint32>{})
+            , lru_module( number_of_sets, number_of_ways) 
+        { 
+            for ( auto& map_of_ways : data)
+                map_of_ways.reserve( number_of_ways);  
+        }
 
-        /* lookup the cache and update LRU info */
+        // lookup the cache and update LRU info
         std::pair<bool, uint32> read( Addr addr);
-        /* find in the cache but do not update LRU info */
+        // find in the cache but do not update LRU info
         std::pair<bool, uint32> read_no_touch( Addr addr) const;
-        /* create new entry in cache */
+        // create new entry in cache
         uint32 write( Addr addr);
 
-        uint32 set( Addr addr) const;
-        Addr tag( Addr addr) const;
+        uint32 set( Addr addr) const 
+        { 
+            return ( ( addr & addr_mask) / line_size) & (number_of_sets - 1); 
+        }
+        Addr tag( Addr addr) const { return ( addr & addr_mask) / line_size; }
+
+    private:
+        const uint32 number_of_sets;
+        const Addr   addr_mask;
+
+        // to convert num_ways to tags
+        std::vector<std::vector<std::pair<bool, Addr>>> ways_to_tags;
```

Comment:
`std::pair` is not the best idea.
1. It is usually unclear what "first" and "second" mean
2. One day we will need to add one more field to the table. With pair/tuple we shall re-write everything!!